### DISCUSSION TITLE
Using "where" in many-to-many relation

### DISCUSSION THREAD

**Discussion by tomas-land** [prisma/prisma#17823] [Q&A]
Trying to figure out how could I get Categories with Expenses created at specific date, any suggestions? Below my prisma schema :

![Capture999](https://user-images.githubusercontent.com/72792707/217638762-08405017-b04c-4baf-b099-87c345581e9a.JPG)

It should look something like this, but "where" is not working in deeply nested relation:

![Capture9999](https://user-images.githubusercontent.com/72792707/217643870-64471909-2727-475b-b932-c2ad3b690ca7.JPG)

**Comment by ludralph** [comment]
Hi @tomas-land 👋 

The `where` argument is invalid for the `select` property in the Prisma query.

One way I would approach this is to use the `prisma.category.findMany` method to retrieve all Category models, including their associated expenses. The resulting categories are then filtered using the JavaScript filter method, checking if any of the associated expenses have a createdAt field that matches the specificDate specified. The resulting categoriesWithExpenses variable will contain an array of Category models, each with its associated expenses created on the specified date.

```js
const specificDate = new Date("2023-02-09T08:32:42.766Z");

const categories = await prisma.category.findMany({
  include: {
    categoriesOnExpenses: {
      include: {
        expense: {
          select: {
            id: true,
            createdAt: true,
            amount: true,
            desc: true
          }
        }
      }
    }
  }
});


const categoriesWithExpenses = categories.filter(category => {
  return category.categoriesOnExpenses.some(coe => {
    return coe.expense.createdAt.getTime() === specificDate.getTime();
  });
});
console.dir(categoriesWithExpenses)
````
output is 
```js
[
  {
    id: 1,
    name: 'Travel',
    createdAt: 2023-02-09T08:32:42.749Z,
    updatedAt: null,
    categoriesOnExpenses: [
      {
        categoryId: 1,
        expenseId: 1,
        assignedAt: 2023-02-09T08:32:42.766Z,
        expense: {
          id: 1,
          createdAt: 2023-02-09T08:32:42.766Z,
          amount: 500,
          desc: 'Airfare'
        }
      }
    ]
  }
]
```

**Reply by tomas-land** [reply]
First of all, I would like to thank you for your solution. I changed my schema to one-to-many relation, because previous was redundant. My goal was to get Categories with Expenses for current month. I searched prisma docs thoroughly , but couldn't find what I need. So I used your advice and filtered Expenses which were created at current month. But I guess it's not the best solution because I always fetching from database all Categories with all Expenses ( including previous months ), so when there will be hundreds expenses it could probably slow down my app. I used "take" option in findMany method, to limit to 150 returned objects, and hoping they will not exceed this number .

**Reply by tomas-land** [reply]
Sorry, I was wrong about limiting, it only limits Categories, but not Expenses .Any suggestions? :)

**Comment by ludralph** [comment]
Hi @tomas-land 👋 

I'm assuming since it's now 1 to many, your schema will look something like this 
```js
model Category {
  id        Int       @id @default(autoincrement())
  name      String
  expenses  Expense[]
}

model Expense {
  id         Int      @id @default(autoincrement())
  amount     Float
  createdAt  DateTime
  categoryId Int
  category   Category @relation(fields: [categoryId], references: [id])
}
```
To fetch only Categories with Expenses for the current month, you can define a custom query using the $where argument to filter the Expenses by createdAt. Here's an example query that shows how to do this:
```js
const categories = await prisma.category.findMany({
  where: {
    expenses: {
      some: {
        createdAt: {
           gt: new Date('2023-02-01T00:00:00.000Z').toISOString(),
          lt: new Date('2023-03-01T00:00:00.000Z').toISOString(),
        }
      }
    }
  },
  include: {
    expenses: true,
  },
  take: 150, // optional limit to the number of returned Categories
});
``` 
The sample result i got is 
```js
[
  {
    id: 1,
    name: 'First Cat',
    expenses: [
      {
        id: 1,
        amount: 500,
        createdAt: 2023-02-14T08:53:08.255Z,
        categoryId: 1
      }
    ]
  },
  {
    id: 2,
    name: 'Secong Cat',
    expenses: [
      {
        id: 2,
        amount: 1000,
        createdAt: 2023-02-14T08:53:27.323Z,
        categoryId: 2
      }
    ]
  },
  {
    id: 3,
    name: 'Third Cat',
    expenses: [
      {
        id: 3,
        amount: 1500,
        createdAt: 2023-02-14T08:53:49.326Z,
        categoryId: 3
      }
    ]
  }
]
```
Let me know if this helps

**Reply by tomas-land** [reply]
Hello again, I already tried this query before, but to make sure I tried again, and the same result. So the problem is that it returns  Categories who have at least one Expense which was created this month and adds from previous month as well. So filtering with ''some" doesn't work as I expect. You see below:

![Captureddfd](https://user-images.githubusercontent.com/72792707/220975915-b9309a87-d416-47a0-a445-329c4d714f27.JPG)

And limiting returned objects with " take" , it limits only Categories, not Expenses. So I quite dissapointed, maybe there is solution but I can't find out, I tried to use raw query ( Prisma supports the option of sending raw queries to  database)  , but not good at raw sql , after several hours tests, I couldn't create this kind of json structure . What do you  think?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
